department-of-veterans-affairs/next-build
# VACMS-16639 broken links

## Description
Closes #[16639](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/16639).
Adds a script for checking broken links. Updates the sitemap to produce indexed sitemaps. Updates some sitemap utilities for re-use across testing (broken links, full a11y scan).


## Testing done
- Tested against both my local (`http://127.0.0.1:8001` & the live https://www.va.gov site.
- Ensured that indexed sitemaps could be parsed to a full list of URLs.


## Screenshots
```
SITE_URL=http://127.0.0.1:8001/ node scripts/check-broken-links.mjs
Number of pages to check: 10172
..............................!.....................................!.........!!.....................................!.............!!!!.........................................!......................................................!.....................................................!.................................................................!.......................................................!................................................!...................................!.!...........................................!.........!!.....................................!..........................................!......................................................!.....................................!............................................ 

[and then a ton more dots]

All batches complete in 23:35!
Scanned total of 423775 links on 10172 pages!
Detected 7104 broken links.

...

https://www.toysfortots.org/Default.aspx
   STATUS: 404
   SOURCE: /dayton-health-care/stories/12-ways-to-create-community-this-holiday-and-connect-with-other-veterans/

https://www.mclnational.org/find-a-detachment.html
   STATUS: 0
   SOURCE: /dayton-health-care/stories/12-ways-to-create-community-this-holiday-and-connect-with-other-veterans/

 Report file written to: /Users/heffner/dev/next-build/broken-link-report.json
```

## QA steps
To test against local output, pull the branch and first run `yarn export` && `yarn build:sitemap` to build the site & sitemaps locally. Then run `yarn export:serve` to start a webserver. You will be testing against that webserver output.

Run `SITE_URL=http://127.0.0.1:8001/ node scripts/check-broken-links.mjs`. It takes about 25 minutes for a full scan on my machine.


To test against the live VA site, run `SITE_URL=https://www.va.gov/ node scripts/check-broken-links.mjs`. Note that the number of pages to check is ~1.5x as large as the next-build current output (10k vs 25k, so the full scan will take a considerable amount of time).


```[tasklist]
```
